### ISSUE TITLE
USB DAC dropouts/glitches

### ISSUE THREAD

**Issue by DBMandrake** [raspberrypi/linux#2215]
Hi,

I'm testing a Behringer U-Phoria UMC204HD USB sound interface with my Pi 2B running the latest version of Volumio and I'm experiencing random glitches/dropouts in the sound which after some investigation I suspect may be related to USB packet loss or timing - issues that have been looked at in the past for the Pi.

The interface supports up to 192Khz 24bit with 4 output channels, however I'm seeing the same symptoms down at 44Khz 16 bit - which is a pop or momentary dropout in the audio at random roughly every 3 to 10 seconds, whether playing audio via Volumio or directly using aplay from the command line. The sample rate makes very little difference to the frequency of occurance of the pops and clicks.

I've already experimented with the various dwc options discussed elsewhere and nothing seems to make much difference - the only one that made some difference was dwc_otg.speed=1 to disable hi-speed mode, which did significantly reduce the frequency of the glitches but did not eliminate them, and of course that is not a viable solution since it affects Ethernet speed etc...

These are the current cmdline options and kernel version used by Volumio that I am testing with:

`splash quiet plymouth.ignore-serial-consoles dwc_otg.lpm_enable=0 dwc_otg.fiq_enable=1 dwc_otg.fiq_fsm_enable=1 dwc_otg.fiq_fsm_mask=0x3 console=serial0,115200 kgdboc=serial0,115200 console=tty1 imgpart=/dev/mmcblk0p2 imgfile=/volumio_current.sqsh elevator=noop rootwait smsc95xx.turbo_mode=N bootdelay=5 logo.nologo vt.global_cursor_default=0 loglevel=0
`

`Linux version 4.9.36-v7+ (dc4@dc4-XPS13-9333) (gcc version 4.9.3 (crosstool-NG crosstool-ng-1.22.0-88-g8460611) ) #1015 SMP Thu Jul 6 16:14:20 BST 2017`

For a comparison I installed the Cubox version of Volumio on a dual core Cubox-i that I have which is a comparable speed to a Pi 3, (so a bit faster than my Pi 2) and the interface works perfectly with both Volumio and aplay with no dropouts whatsoever, so I think this rules out a problem with the interface being supported properly in Linux on arm devices.

Here is the output from lsusb:

[lsusb.txt](https://github.com/raspberrypi/linux/files/1346360/lsusb.txt)

If there is anything else I can do to test or help debug this please let me know.

**Comment by DBMandrake** [comment]
After further reading I think I may be affected by the problem described in this issue:

https://github.com/raspberrypi/linux/issues/2134

I will attempt to build a custom version of Volumio with the very latest kernel (rpi-update cannot be used) to see if it fixes the problem as even the latest dev build of volumio is only 4.9.41 - shown to be bad in issue 2134.

**Comment by DBMandrake** [comment]
I've now tested a 4.9.51-v7+ kernel hoping that the timer fix for isochronous transfer in 43 described above would be the solution but unfortunately it doesn't fix this issue.

The severity of the problem seems to be reduced by about half - with a 36 or 41 kernel I get on about 6 dropouts per minute which audibly last probably 1/10th of a second, with the 51 kernel I am seeing maybe 3 dropouts per minute and they don't "sound" as bad, but they are definitely still there and the audio is not bit perfect or glitch free, and can easily be heard on a sinewave test signal played using aplay.

Is there any further testing or debugging I can do to get to the bottom of this issue or at least prove whether it is a hardware limitation or still potentially fixable in software ?

**Comment by DBMandrake** [comment]
More testing and some interesting conclusions. First of all this problem is not fixed by newer kernels and doesn't seem to be related to the timer fix in issue #2134, unfortunately. I can reproduce the same problem with 4.9.36, 4.9.41 and 4.9.51 with very little difference contrary to what I said above.

I found that when I tested on a fresh install of Raspbian I was not experiencing audio glitches using aplay. at either 44.1Khz 16 bit or 96Khz 16 bit.

I then tried to narrow down what was different between Raspbian and Volumio. It turns out that the main Volumio service - which is a node js server for the web interface, when running leads to glitches/dropouts in the audio. When I stop that service the glitches almost entirely stop although I can still reproduce a very occasional glitch with it stopped, maybe once every 2-3 minutes, instead of every few seconds when it is running.

Nodejs is using very little CPU time (under 10%) when the web interface is idle yet still causes pretty bad sound glitches. If I renice nodejs to a low priority or aplay to a high priority it makes no difference at all, which suggests its not aplay being starved for CPU time. (aplay never reports any underruns anyway)

I experimented with the snd-usb-audio nrpacks module option - no change at all. I also experimented with every buffer related alsa configuration setting - no improvement. Finally I tried dwc_otg.speed=1 to disable USB2.0 support and the symptoms are gone - absolutely no glitches or dropouts in about an hour of testing on test signals and music, despite nodejs running as before.

So something the nodejs service is doing is causing issues with USB2.0 transmission - perhaps certain syscalls that it is calling are disabling interrupts for a long time, or the syscalls are just taking too long to return leading to the kernel missing a time critical USB transmission.

Perhaps a realtime kernel would help here. I would add that this sound interface is USB2.0 (but also supports fallback to USB1.1) so I don't think this relates to the fiq fixes from the last few years ?

Is there anything I can do to improve the USB 2.0 performance to avoid dropouts in the audio or am I stuck using USB1.1 as a workaround ?

**Comment by P33M** [comment]
Hello.

From an examination of your lsusb listing, the audio device is a high-speed device.  The descriptors for the isochronous OUT endpoints (i.e. audio streams) also specify a bInterval of 1 which means a packet transfer every microframe - this is a particularly strenuous requirement for the Pi as it means the maximum hardware interrupt latency tolerable is <125uS.

The FIQ code attempts to improve matters by performing batches of transfers, but it can only perform as many transfers as the submitted URB requests - at the boundary where URBs are returned/the next one queued, control is passed to the IRQ-driven driver so there's a window of vulnerability where a latency spike could cause frame slippage.

dwc_otg.speed=1 will increase the hardIRQ latency tolerance to 1ms as that's the frame interval at full-speed. It's no surprise that the glitches disappear when clobbering the bus speed.

The issue described above should cause at most momentary audio dropouts (typically less than 1ms) but this is a function of how the USB device behaves when presented with an underrun condition (i.e. missing data). Can you capture a typical output stream via line-out -> recording device (e.g. line-in on a PC or similar) and upload the wav somewhere?

What's puzzling is the apparent difference between running a "heavy" userspace application (which should have negligible effect on hardIRQ latency) and plain aplay.

**Comment by popcornmix** [comment]
@P33M one thing @DBMandrake mentioned to me was strace showed node js doing a number of cacheflush operations. It does make use of JIT and so is likely triggering instruction/data cache flushes in the kernel. 
I wonder whether there is any disabling of interrupts when that occurs? (or otherwise harming of interrupt latency)

**Comment by DBMandrake** [comment]
> From an examination of your lsusb listing, the audio device is a high-speed device. The descriptors for the isochronous OUT endpoints (i.e. audio streams) also specify a bInterval of 1 which means a packet transfer every microframe - this is a particularly strenuous requirement for the Pi as it means the maximum hardware interrupt latency tolerable is <125uS.

That makes sense. This interface is designed for home recording studio use and as such minimising round trip latency from record to playback is a key factor in it's design, possibly at the expense of not being as "easy" to drive by the host OS.

The problem felt like a latency to me issue as well - the glitches don't seem particularly worse if I play 96Khz 24bit (which I believe is sent as 32bit to the card) than if I send 44.1Khz 16 bit, suggesting it wasn't a throughput issue.

> The issue described above should cause at most momentary audio dropouts (typically less than 1ms) but this is a function of how the USB device behaves when presented with an underrun condition (i.e. missing data). Can you capture a typical output stream via line-out -> recording device (e.g. line-in on a PC or similar) and upload the wav somewhere?

Yes I can can make some recordings for you, I should be able to do this tomorrow. The severity varies from a small "click" that you can only hear on a simple and uncluttered test signal like a sine wave (but not easily in music) all the way to a loud "pop" that is easily heard during music.

Even if the click was subtle though, any glitch or dropout does defeat the purpose of trying to set up bit perfect playback from a high quality DAC.

>What's puzzling is the apparent difference between running a "heavy" userspace application (which should have negligible effect on hardIRQ latency) and plain aplay.

Yes it is very puzzling - when I identified nodejs as the application that was placing the problematic "load" on the system (stopping the service eliminated 99% of the symptoms) I spent a few hours trying to figure out why.

I found it didn't seem to be just raw CPU use that was causing the issue, as adjusting relative priority of the nodejs service (which is a parent thread and 3-4 worker threads) and aplay made absolutely no difference, and I was still getting glitches even when nodejs was mostly "idle". (less than 5% CPU use) 

Also other high cpu loads that I tried putting on the system were not causing trouble. Memory use also wasn't an issue as the service is using about 200MB of ram and the system is not using swap and has plenty of free ram.

I concluded that if it was causing this problem with such low cpu use even at low scheduling priority it must be making system call(s) that were somehow harming interrupt response time so I decided to strace the parent nodejs process and as popcornmix says I noticed there were a lot of cacheflush calls being made, especially soon after the service was started. (When audio glitches were even worse still)

The cacheflush calls might be a red herring, but it was the only thing that jumped out at me and seemed like something worth following up.

It's possible the problem isn't even in the USB driver at all but somewhere else in the kernel where some user space accessible system call(s) are doing something that is hurting interrupt response times or maybe even causing driver interrupts to be missed altogether ?

**Comment by DBMandrake** [comment]
As promised here are some recordings of the problem made using audacity via line input, they are in 192Kbit 32 bit float wav format, uncompressed. This is a 20Hz sinewave - low enough in frequency that it won't be heard on most speakers/headphones but the resulting clicks will be.

The first one was made after volumio had been booted for a while and was largely "idle". cpu use of nodejs at this point is under 10%:

https://drive.google.com/open?id=0B5qpN5dK9MKjdWc4Y2F2SmlDNTA

Clicking is more subtle and I count roughly 6 glitches over 30 seconds. I measured a couple in Audacity and they were about 0.4ms long, but because the amplitude jumps back to near zero it makes it quite audible.

This second recording was started while the nodejs service was stopped, and then the service was started at about 5 seconds into the recording, with symptoms much more severe as a lot more work is being done by the service including setting up child threads:

https://drive.google.com/open?id=0B5qpN5dK9MKjOURST3VfWUdEVHM

It takes a few seconds but it starts crackling badly between 16 and 22 seconds. I measured one dropout at about 17.7 seconds that lasts a full 8ms, which suggests something is very wrong.

Here is the original test file I'm playing with aplay:

https://drive.google.com/open?id=0B5qpN5dK9MKjbkhhdTFfdHNzbDg

Most of the dropouts appear to be the sample simply dropping to near zero for the duration then returning to where it should be, but I did see at least one dropout in the second file where you can see from the sinewave that time sync has been lost altogether and the sample has been "delayed", at about 17.79 seconds.

Just to try to rule out cpu hogging I repeated the test again with the nodejs service niced to 19 and it was no better, if anything it was worse: (although there is a lot of random variation from test to test)

https://drive.google.com/open?id=0B5qpN5dK9MKjTXh4Nk5YWGxCMkE

In this last one I see a 3ms dropout where the time sync seems to have slipped at 16.38s although most of the glitches are about 0.5ms in length. Here is a screen grab from that area:

![image](https://user-images.githubusercontent.com/6304155/31381458-7d620776-adac-11e7-8d45-b2bdfa32f210.png)

Hopefully these files are useful to help at least classify the nature of the problem - if you need any more testing done please let me know.

**Comment by musicwonder** [comment]
Hi there, we are happy to find this thread. In short: We have the same problem with the ARM architecture. We tried:
- Hardware: Raspberry Pi 3, Asus Tinker Board
- Software: Raspbian, Volumio, Armbian, Ubuntu Mate

All the hints we found in lots of forums didn't help, but
It WORKS on:
- Hardware: UP Board with Atom Processor, PC Mainboards
- Software: Linux Mint

So it's clear that there must be a timing / time slice issue with the isochronous data streams on ARM devices.

**Comment by P33M** [comment]
@musicwonder Can you confirm which USB controller the DAC is connected to on the tinker board?

A `lsusb -t` will tell you.

**Comment by DBMandrake** [comment]
>So it's clear that there must be a timing / time slice issue with the isochronous data streams on ARM devices.

It's not actually all arm devices. While this problem is being looked into I've been running Volumio on a dual core Cubox-i (imx6 arm architecture) and I have absolutely zero problems with glitches or dropouts with this USB interface even when the box is under heavy load. Unfortunately the Cubox-i build of Volumio is older than the Pi build and not kept as up to date by the developers (new versions are always released for the Pi first) so I would prefer to run the Pi version if the problem can be solved satisfactorily.

I've also tried this Behringer interface on a Vero4k (Amlogic S905X arm chipset) and there are no glitches there either, although to be fair this is much faster hardware than a Pi so is not a fair comparison, and there are no Volumio builds available for it. (I tested with both Kodi and aplay)

**Comment by musicwonder** [comment]
@P33M: It happens at each of the 4 USB Ports.

**Comment by P33M** [comment]
I've been toying with the idea of adding low-key profiling to the dwc_otg driver - it's possible to get reliable indication that you're missing interrupts (FIQ->IRQ latency being too big, for example). 

Unfortunately time is tight at the moment so it will be a while before I can look at this.

**Comment by musicwonder** [comment]
Is there a chance that the issue is gone with a rt-kernel? If so, I would it give a try.

**Comment by DBMandrake** [comment]
>Unfortunately time is tight at the moment so it will be a while before I can look at this.

No problem - meanwhile I am using it on my Cubox, when you get a chance to look at it I'll be around to do any testing required.

**Comment by musicwonder** [comment]
No I have made a test with an orange pi pc2 with armbian 4.13. Same issue but less often. One dropout per 1-4 minutes. Why only arm devices?

**Comment by musicwonder** [comment]
rt-kernel doesn't help. Tested.

**Comment by musicwonder** [comment]
Update: With orange pi pc2 and a new install of armbian 4.13 the sound is about one minute clear. After that a first drop out comes. Then I have many glitches and dropouts and buffer underrun while using brutefir (2 In/4 Out with an USB c-media soundcard). Reboot doesn't help. You cannot get the same clear sound how it is after a new installation. Could there be an issue with SD Card read/write?

**Comment by JamesH65** [comment]
Has there been any further progress on this issue?

**Comment by DBMandrake** [comment]
There's nothing further I can do at the moment but I'm able to test any proposed fixes.

**Comment by volumio** [comment]
I've been banging my head quite a lot on this lately. In my case disabling nodejs made things a bit better, but dropouts were still there.
Those are my results so far:
- Was able to reduce significantly (almost none) dropouts by setting MPD cpu affinity to a core and setting its realtime priority (with RT kernel) to 55. However this made the whole system unstable.
- Played with lot of possible combinations of fiq\fsm values. The only thing that worked a bit was to set those params in cmdline.txt:
dwc_otg.fiq_enable=1 dwc_otg.fiq_fsm_enable=1 dwc_otg.fiq_fsm_mask=0xF dwc_otg.nak_holdoff=1
However I still get occasional dropouts, and I'm not sure if the dwc_otg.nak_holdoff value makes any sense.
Any pointers will be appreciated, I can test proposed fixes and recompile the kernel if needed.

**Comment by DBMandrake** [comment]
>Any pointers will be appreciated, I can test proposed fixes and recompile the kernel if needed.

Did you try dwc_otg.speed=1 ? For me this almost eliminates the issue. Of course this is only a workaround, and restricts the USB controller to USB 1.1 mode and therefore limits all USB devices (including the onboard network adaptor) to 12Mbps.

However if your only USB devices are the sound card and the onboard Ethernet controller and you are not trying to use it for anything other than Volumio then this seems to be fast enough - after all audio doesn't require very high bitrates.

As P33M said earlier this reduces the required interrupt response time from 125uS to 1mS, making the interrupt load much easier to deal with.

Might be a usable workaround for you in the meantime. I've sidestepped the issue by running Volumio on a Cubox-i for now but that introduces other issues like the fact that the Volumio builds for that device are very out of date so I'd like to come back to the Pi when I can.

>Was able to reduce significantly (almost none) dropouts by setting MPD cpu affinity to a core and setting its realtime priority (with RT kernel) to 55. However this made the whole system unstable.

I see the same glitches testing with aplay from the command line and also airplay (shairport) when mpd isn't even the active player process, so the issue is a lot more deep seated than that and seems to be down at the interrupt handler/scheduler level in the kernel.

**Comment by volumio** [comment]
> Did you try dwc_otg.speed=1 ? For me this almost eliminates the issue. Of course this is only a workaround, and restricts the USB controller to USB 1.1 mode and therefore limits all USB devices (including the onboard network adaptor) to 12Mbps.

I would like to avoid this, as it's not really a solution, since it will mean max 24/96 and it will make db indexing very slow.

**Comment by P33M** [comment]
>@P33M one thing @DBMandrake mentioned to me was strace showed node js doing a number of cacheflush operations. It does make use of JIT and so is likely triggering instruction/data cache flushes in the kernel.
>I wonder whether there is any disabling of interrupts when that occurs? (or otherwise harming of interrupt latency)

The ARM SVC handler explicitly enables interrupts before calling whatever ARM-specific handler is needed:  http://elixir.free-electrons.com/linux/latest/source/arch/arm/kernel/entry-common.S#L172

So the cache maintenance operation should be done without hardirqs disabled.

The coherent_user_range() function invalidates Icache and cleans Dcache out to the inner shareable boundary which I imagine takes many cycles, but there shouldn't be anything like holding off interrupts for dozens of microseconds in that operation...

**Comment by volumio** [comment]
@popcornmix @P33M 

are there in your opinion some kernel confs or userspace tweaks that can be applied in this particular scenario (Hi RES USB Music playback) that you think can mitigate\solve this issue?

**Comment by JLPER** [comment]
Hi all

I would like to inform that I have exactly the same problem with differents setups connected to asynchronous USB DAC XMOS WAVEIO, whatever the music format
- BeagleBone Black with Volumio 2.194
- RPI B+ v1.2 with Volumio 2.348 or Moode Audio 

I have posted some logs here [ http://logs.volumio.org/volumio/mD6JT8I.html ](url)

Good luck for debugging it really is annoying.

BR

**Comment by nyebi** [comment]
Although i'm not hundred percent sure i think i might be experiencing the same problem with a Creative X-Fi Surround 5.1 pro USB DAC.
Playing back something in Kodi (post 7.0.3 LibreELEC builds with a kernel newer than 4.4.13 and also on the latest raspbian) that has an audio track with more than 4 channels, eg 5.1 whenever there is silence i start to hear a crackling noise.
I was not able to reproduce the issue using VLC on the same system.
Reverting back to an older kernel (not sure about the exact version but 4.4.13 seems to be better) the issue remains but the noise becomes so quiet that i am barely able to hear it.
lsusb -v and samples can be found here: https://1drv.ms/f/s!Ask92-8JR3iXhAp25gE0TisGy5Qw
The first sample is recorded on LibreELEC 8.2.2, the second on LibreELEC 7.0.3. In the second sample the clicks are barely audible but they are worse IRL.

**Comment by mr-berndt** [comment]
I am fighting with this issue for a long time already.

Had it with a LFS system with initd and found the issue not present using the same hardware with raspbian. The main difference seemed that raspbian used systemd and cgroups which I didn't have and could not test.

I switched over to a buildroot system with systemd and cgroups but the problem remains. I looked for any special configuration of cgroups in raspbian but could not find anything.

Kernel-configs an both mentioned systems were always the default targets of which I understand raspbian is using as well. - Are raspbian kernels patched in any way?

**Comment by gvdw** [comment]
@DBMandrake your audio glitches sound exactly what I encounter on my PC using a basic USB audio DAC plugged into USB 2 port - random clusters of glitched audio (usually about 30 seconds of clear audio then it starts glitching).  So this issue is not specific to raspberry pi or ARM and the comment by @musicwonder that it worked on one x86 platform therefore it's ARM-specific isn't right.

Pre-emptive kernels or ones running a faster kernel timer do not help.  I even have a BIOS option for Isochronous support and flipping that doesn't change anything. Using usbmon I can see there is a very slightly different message passed only at the exact time of the crackle, but I was not able to figure out what the messages meant, other than it LOOKS like a status flag showing a negative number rather than zero (in crackle-free periods).

What helped on my PC was plugging my DAC into the USB3 bus, rather than USB2.  Unfortunately I haven't diagnosed it further, once I figured out a workaround I gave up.  Unfortunately on the Pi you only have 1 root port.  You could try an older Raspbian with a 3.x kernel to see if it is any different? (I have been using 4.4, 4.13, 4.15 on my PC).

**Comment by DBMandrake** [comment]
@gvdw

>@DBMandrake your audio glitches sound exactly what I encounter on my PC using a basic USB audio DAC plugged into USB 2 port - random clusters of glitched audio (usually about 30 seconds of clear audio then it starts glitching). So this issue is not specific to raspberry pi or ARM and the comment by @musicwonder that it worked on one x86 platform therefore it's ARM-specific isn't right.

As I mentioned, not all arm devices are affected. I am running Volumio on a Cubox-i without any glitches at all using the same sound interface.

Having said that, I've recently realised that my current version of Volumio on Cubox-i is using a much older 3.14 kernel. One of the Volumio devs is working on a Cubox-i build based on a 4.x kernel and, surprise surprise, is seeing problems with glitches and dropouts with USB audio.

Which makes me wonder, is this a more general problem with USB Audio in 4.x kernels and the only reason my Cubox-i is glitch free is because it is still running 3.14 ?

It would be interesting to test a Pi on an old 3.14 kernel however this is becoming difficult as many of the newer Pi models were released after the 3.x kernel series was no longer being used so a 3.x kernel could only be booted on old Pi hardware. Might be worth someone trying though.

>Pre-emptive kernels or ones running a faster kernel timer do not help. I even have a BIOS option for Isochronous support and flipping that doesn't change anything. Using usbmon I can see there is a very slightly different message passed only at the exact time of the crackle, but I was not able to figure out what the messages meant, other than it LOOKS like a status flag showing a negative number rather than zero (in crackle-free periods).

What Linux distro are you running ? Is there any way you could try compiling an old 3.14 kernel for it to see if you still have glitches/dropouts ? That may be difficult with a modern distro especially if it is running systemd, but also might be worth a try.

**Comment by JLPER** [comment]
Hi
I have tried several mpd setup including Volumio Rune and Moode on Pi and BBB and at one moment the sound got worse on theses glitches with same USB device. So I subscribe to the fact that from a moment the version of OS was not okay. But I have no track of the precise moment/version it happened.
BR

**Comment by gvdw** [comment]
> What Linux distro are you running ?

@DBMandrake I'm on Mint 18.3, I tried a 3.19 kernel and a 3.14 and both still caused me dropouts via USB2.

For anyone who's interested in digging into USB transfers see [here](https://pastebin.com/raw/5fmiybPd) for the usbmon output - look for the "-63" on the middle line it seems to be a status flag in the callback from the DAC indicating something went wrong.

**Comment by moritzvieli** [comment]
I have exactly the same problem. Except, gstreamer is responsible for the pops (this pipe gst-launch-1.0 --gst-debug=3 playbin uri=file:///test/vid.mp4) and not node. I also checked it with strace, but did not find many cacheflushes. CPU is also not really used. I suspect it's the USB bus.

**Comment by gvdw** [comment]
I have worked out on my PC that it only crackles when I have software playthrough enabled (meaning 1x isochronous transfer in each direction, to monitor a recording).  With just the recording transfer and no playback going, I get no dropouts.

Since the Pi has Ethernet and USB sharing the same controller it may be a problem of limited resources

**Comment by moritzvieli** [comment]
Ok, can confirm the problem is related to the USB (driver). I bought a cheap new USB audio interface which has a bInterval of 10 instead of 1 and it works flawlessly now. This really seems to be an issue. On Windows, I can configure a "USB buffer" with my old interface. Is this somehow also possible on Linux? Can I somehow increase this bInterval without writing an own driver? :)

**Comment by jsc-paneda** [comment]
I've been using an XMOS based USB sound card and Jack (with RT) on the host.
I have noticed crackles also on x86. On x86 it only crackles when I run the card full duplex (capture + play simultaneously). The crackles on x86 are are not very frequent and may not appear at all.
On Rpi3 I get crackles more frequently and also for single directions.
So my conclusion is that it's a general alsa/usb problem and does not depend on RPi:s USB/Ethernet sharing controller. Of course the use of the sharing controller may trigger the problem to occur more frequently.

**Comment by Wang-Yue** [comment]
Is there an upstream bug that I can track?

**Comment by pelwell** [comment]
I've not found one yet.

**Comment by Wang-Yue** [comment]
@pelwell,  You mentioned that "There is some hope that the glitching is due to a specific change in the handling of softirqs - Linus is aware and they are working on a fix." so I assume there should be an upstream bug report? If not, please create one

**Comment by pelwell** [comment]
Let me rephrase that: In the few minutes I spent trying to find the Linux mailing list conversation, I failed to find it. Perhaps you have more time, and will have more luck finding it. If you do, please report back here.

**Comment by Wang-Yue** [comment]
I'm not able to find one too, and that's why I was asking. 
I think it's better to create a new bug report, otherwise this bug would never get fixed.

**Comment by pelwell** [comment]
I've refound a thread on the subject: https://patchwork.kernel.org/patch/10150457/

The next step is to find softirq patches in newer kernels, in case it has already been fixed. But this bug is big enough that it is not going to be forgotten.

**Comment by Wang-Yue** [comment]
I'm not sure if it ever get fixed. As far as I tested the issue is still present in 4.19.17 (I update via rpi-update using the next branch).

**Comment by Wang-Yue** [comment]
@volumio  Just tested what you tried --- with real time kernel and set cpu affinity and priority to 55. Still a lot of pops, especially with a DSD512 test file. On my low end PC things run very smoothly.

just curious, how does Volumio Primo with a tinkerboard solve this issue? Or does it just not appear when using i2s?

**Comment by JLPER** [comment]
Yes in I2s no problem for me though i cannot use my WaveIO usb gateway

**Comment by alastair-dm** [comment]
> I've refound a thread on the subject: https://patchwork.kernel.org/patch/10150457/

That thread is about [this commit](https://github.com/raspberrypi/linux/commit/3eb86ff32eb54c4345b723ae8dd03bd7487d35bd) so it's been in newer kernels for a while.

**Comment by Wang-Yue** [comment]
@alastair-dm I tried version with the commit you mentioned but the problem is not solved.

**Comment by alastair-dm** [comment]
Sorry, I should have been clearer. Based on that commit it appears to me that the patch has been in the Raspbian kernel for a while, so anything above 4.14.64 should include it. I think Volumio tracks the Raspbian kernel, so should be including the patch too, but the audio glitches persist under certain circumstances. I haven't built a kernel from source to be completely certain though.

The commit is also useful in pointing out the timeline for softirq changes, and the specific commits involved. If we can be clear whether the initial change in 4.9 started the problem we'll have a solid basis for an upstream bug report.

**Comment by Wang-Yue** [comment]
@alastair-dm I have a DSD512 file and a DAC that supports DSD512 and can reproduce the issue very easily --- when it occurs the glitch is very evident. If there's a easy way I can roll back to previous versions of kernel images I can do a binary search and find when did the problem occurred.

**Comment by alastair-dm** [comment]
@Wang-Yue if I knew an easy way I'd be testing it myself ;-) If you can find old kernel packages that would probably be easiest, but I suspect you'll need to check out source from git and compile your own.

**Comment by Wang-Yue** [comment]
@alastair-dm compiled binaries could be found in https://github.com/Hexxeh/rpi-firmware (which is a mirror of the official compiled kernel)

essentially you just need to do a binary search on the commit history and copy the kernel7.img to /boot and modules files to /lib. 

the script https://github.com/Hexxeh/rpi-update may make the process easier but I didn't look into it closely.

**Comment by DBMandrake** [comment]
I don’t think rpi-update is compatible with volumio.

**Comment by jsc-paneda** [comment]
I'm developing a USB sound device based on XMOS and I've spent the last couple of weeks trying to understand why these glitches occurs.
I don't have a solution but perhaps an explanation..

In the audio device firmware I've added a measurement of the time between the host's USB tokens. With HS-USB they should appear at the device with a 125us interval.

As the XMOS device is a deterministic real-time system, I expect measurements to be very accurate (+/- 20ns or so).

What I found is that it seems that some in and out "tokens" are omitted (a usb token is sent by the host for every in or out data packet).
I sometimes measure (pretty exactly) 250us between two successive token's. Sometimes I measure (again exactly) 375us between two successive tokens's.
So as the time between some successive tokens's seems to be a multiple of 125us (> 125us) I can only conclude that the host simply omits to issue some tokens's.

As a missing token cannot be repaired (USB audio standard only allows add/remove of a single sample in successive transfers to adjust for time domain differences), this eventually cause the device buffer to overflow and an audible glitch to occur.

I also wrote an isolated test for isochronous transfer in general using libusb (i.e. Alsa/snd_usb_audio is not used at all). 
In this test the device only return an increasing counter value in each iso transfer.

The same issue is present in this test too. So I conclude that its not an Alsa issue but rather a lower level problem with isochronous transfer in general.

I have also tried to increase the number of URB's (=libusb transfers) and also the number of SOF's per URB. It does have some impact but does not solve the problem.

I have used both RPi and x86 as test hosts. I only find this issue when the host is on RPi, not on x86. So I'm pretty certain its not an issue with my device hw.

I only tested one direction so far "USB in" (device to host).

**Comment by P33M** [comment]
That's interesting.
Do the missing SOF events only happen on a Raspberry Pi? Also, as it looks like your XMOS device is a high-speed device, do you get missing SOF events when plugged into an A+ model (which doesn't have a USB hub)?. It would be good to rule out the onboard hub dropping SOF packets.

**Comment by jsc-paneda** [comment]
With my test app I only got missing events on RPi, not when the test host runs on my x86 workstation. <del>However I have previously with Alsa also heard some dropouts on x86 which I suspect was the same issue. In any case on x86 its much harder to get the issue.</del> Edit: The issue is not present on x86, see below..

 I have not tried an A+ model but I will try to get hold on one a make a test.

**Comment by larschristensen** [comment]
> I'm developing a USB sound device based on XMOS and I've spent the last couple of weeks trying to understand why these glitches occurs.
> I don't have a solution but perhaps an explanation..
> 
> In the audio device firmware I've added a measurement of the time between the host's SOF's. With USB-HS they should appear at the device with a 125us interval.
> 
> As the XMOS device is a deterministic real-time system, I expect measurements to be very accurate (+/- 20ns or so).
> 
> What I found is that it seems that some SOF's are omitted.
> I sometimes measure (pretty exactly) 250us between two successive SOF's. Sometimes I measure (again exactly) 375us between two successive SOF's.
> So as the time between some successive SOF's seems to be a multiple of 125us (> 125us) I can only conclude that the host simply omits to issue some SOF's.
> 
> As a missing SOF cannot be repaired (USB audio standard only allows add/remove of a single sample in successive transfers to adjust for time domain differences), this eventually cause the device buffer to overflow and an audible glitch to occur.
> 
> I also wrote an isolated test for isochronous transfer in general using libusb (i.e. Alsa/snd_usb_audio is not used at all).
> In this test the device only return an increasing counter value in each iso transfer.
> 
> The same issue is present in this test too. So I conclude that its not an Alsa issue but rather a lower level problem with isochronous transfer in general.
> 
> I have also tried to increase the number of URB's (=libusb transfers) and also the number of SOF's per URB. It does have some impact but does not solve the problem.
> 
> I have used both RPi and x86 as test hosts. I only find this issue when the host is on RPi, not on x86. So I'm pretty certain its not an issue with my device hw.
> 
> I only tested one direction so far "USB in" (device to host).
> (I still suspect that the issue may be present also on x86 however with a much less noticable frequency when using duplex transfers).

Interesting findings. I have also been experiencing similar dropouts when using a USBStreamer (XMOS-based) connected to an RPI3 in full-duplex mode. In the end, I have managed to get a stable full-duplex audio stream by simply inserting a USB hub between the RPI3 and the USBStreamer. Without the USB hub dropouts will occur, which to me indicates that it must somehow be a low-level USB issue, maybe interrupt related.

**Comment by jsc-paneda** [comment]
I've now made more tests with various hosts.

Short version:
Isochronous transfer mode in general is not reliable on Raspberry Pi.

Long version:
I've checked for missing token events at the device using these hosts:

PC: x86/I7 (kernel 4.20.12) -> No missing token events detected.
PC: x86/Celeron (kernel 4.20.12) -> No missing token events detected.
Tinkerboard: Arm/Rockchip RK3288 (kernel 4.4) -> No missing token events detected.

RPi 3 Model B: (kernel 4.19.25) -> Missing token events detected.
RPi 3 Model B+:  (kernel 4.19.25) -> Missing token events detected.
RPi 3 Model A+:  (kernel 4.19.25) -> Missing token events detected.

I.e. the use of a USB hub chip (LAN9514/LAN7515 etc.) is _not_ causing the missing token events.
Also the A+ model connected directly to the audio device has the same issues.

I have also tested using a higher "bInterval" value, i.e. in addition to using bInterval=1 -> 125us, I also
tested bInterval=4 -> 1 ms. This does not help. I still get missing token events, now with a 1ms diff between successive tokens's.

My conclusions:
According to ([RPi docs](https://www.raspberrypi.org/documentation/hardware/raspberrypi/usb/README.md)) the USB host port is software OTG port, not a true hw USB controller.

Personally, before starting this investigation, I was not aware of this.

Some quotes:
"..In essence, the OTG hardware is simpler than the equivalent hardware on a PC"
"..but to provide an adequate level of functionality for most of the USB devices that one might plug into a Pi, the system software has to do more work."

To me this is likely the cause of this issues. Using a "software" host controller makes isochronous transfer mode unreliable compared to the other hosts I tested.

Hopefully there are ways to improve the OTG controller in RPi for isochronous transfers..

**Comment by P33M** [comment]
So the SOF missing correlates with the isochronous packet interval. In the case where you miss a SOF, do you
a) receive an isochronous packet in the microframe immediately before and
b) is this iso packet correct or corrupt?

The controller hardware does the packet scheduling. If it's failing to issue SOF packets, any software workarounds are likely to be ugly.

**Comment by alastair-dm** [comment]
@jsc-paneda any chance you can test with RPi 3 and kernel <4.9? If you still get missing SOF events we can rule out the softirq changes as the cause.

**Comment by jsc-paneda** [comment]
> So the SOF missing correlates with the isochronous packet interval. In the case where you miss a SOF, do you
> a) receive an isochronous packet in the microframe immediately before and
> b) is this iso packet correct or corrupt?
> 
> The controller hardware does the packet scheduling. If it's failing to issue SOF packets, any software workarounds are likely to be ugly.

In my testing I have no knowledge of what's happening on physical level. I refer to "missing token events". However this terminology might be wrong (I have limited knowledge about physical level USB).

What I do detect is that on top of the XMOS provided USB device library, there are two methods that are called every 125us (or 1ms depending on bInterval). One method for "out" transfers and one method for "in" transfers. For each method respectively, I measure the time between successively calls by the underlying device library.

Its clear that whenever the host is a RPi, some calls are omitted by the underlying device library and I refer to this as "missing token events".

**Comment by jsc-paneda** [comment]
> @jsc-paneda any chance you can test with RPi 3 and kernel <4.9? If you still get missing SOF events we can rule out the softirq changes as the cause.

I have a sd card with older kernel 4.4.41. I do get a different result!

The missing "token events" are almost gone! I ran the test for ~30minutes. I only lost 2 events. This is usually what I loose per minute with the 4.19 kernel.

However!? I get much more audible dropouts with 4.4, roughly one audible glitch every 10 sec.
With the 4.19 kernel I only get audible glitches when I also detect a missing event.

This is a very different result indeed.

**Comment by DBMandrake** [comment]
> I have a sd card with older kernel 4.4.41. I do get a different result!

Careful, the symptoms of this problem vary dramatically depending on CPU (system call ?) load. In the case of Volumio for example there is something about the node.js server that it runs that seems prone to trigger this issue particularly badly as if that service is stopped most of the audio glitches stop. (But not completely)

So if it's a completely different installation on a different SD card the comparison may not be valid depending on what userspace software is running and perhaps even depending on SD card write performance if slow disk writes are causing delays in the kernel.

Ideally you would swap kernels on the same installation on the same SD card without introducing any other uncontrolled variables.

**Comment by jsc-paneda** [comment]
I agree and I should probably clarify that all testing on the 4.19 kernel was done with the same sd card swapped into different RPi:s. The 4.4 is a different sd card.

However both sd cards are pretty much clean ArchLinux installations with no custom services running.
The host test command is simply:
arecord -Dplug:hw:0 -f s16_le -c2 -r48000 | aplay -Dplug:hw:0 -f s16_le -c2 -r48000
..and that's pretty much all that is running in userspace.

I then listen to a sine tone the on board audio monitors while looking for missing events.

**Comment by JLPER** [comment]
Hi all
To be confirmed by someone but I did have the same issues with Moode Audio that I stopped using for that reason.
Bests

**Comment by alastair-dm** [comment]
How easily can you put a self-built kernel on Arch for the Pi? I've been cross-compiling for Raspbian, and it's only a few lines need changing in kernel/softirq.c to undo the changes from 4.9. The problem I have is reproducing the issue in Raspbian but I've been looking at playback only, so I'll give your method a try. I can reliably reproduce in Volumio with brutefir running, but it's not easy to change the kernel in that case.

**Comment by jsc-paneda** [comment]
Can do. I'll have a look next week. It would be great if you could refer me to the commit in question.

**Comment by alastair-dm** [comment]
> Can do. I'll have a look next week. It would be great if you could refer me to the commit in question.

The change in 4.9 is [here](https://github.com/raspberrypi/linux/commit/4cd13c21b207e80ddb1144c576500098f2d5f882) and the later (4.14.64?) change to try to fix some of the problems it caused is [here](https://github.com/raspberrypi/linux/commit/3eb86ff32eb54c4345b723ae8dd03bd7487d35bd). The comments in the former explain what it was fixing, and in the latter the things that broke. It should be sufficient to remove the calls to ksoftirqd_running() added in the 4.9 change, and the kernel then compiles (with a warning about the now unused ksoftirqd_running()). I haven't tried it yet as I can't yet reliably reproduce the issue on Raspbian. Your arecord|aplay method doesn't give any audible issue with my UCA202, same as with volumio and brutefir. The Forte's proprietary control interface isn't supported in linux so I can't use that method with it, so I'll try with brutefir again.

**Comment by Wang-Yue** [comment]
> However!? I get much more audible dropouts with 4.4, roughly one audible glitch every 10 sec.
> With the 4.19 kernel I only get audible glitches when I also detect a missing event.

Can you check what's the cpu usage of 4.4? Maybe with another system you are doing different things (for instance, mpd may do format/sample rate conversion and throttle the CPU) Please make sure you are using the same MPD/ALSA setup and the same audio file.

**Comment by Wang-Yue** [comment]
> Tinkerboard: Arm/Rockchip RK3288 (kernel 4.4) -> No missing SOF events detected.

There're reports above that Tinkerboard has the same issue as raspberry pi.

**Comment by jsc-paneda** [comment]
> > Tinkerboard: Arm/Rockchip RK3288 (kernel 4.4) -> No missing SOF events detected.
> 
> There're reports above that Tinkerboard has the same issue as raspberry pi.

I also get audible dropouts with the TinkerBoard (4.4 kernel). However they are not caused by missing token events.
I also got audible dropouts with my RPi 4.4 test that did not relate to missing token events (see above).

So there are at least two causes for audible dropouts. One cause seems exclusive to RPi hosts (missing token events).

The other cause I have not started looking into yet (I'm not even sure its present in the 4.19 kernel). If present, I have a hunch that it may relate to the time domain adjustment algorithm (defined by the UAC2 standard). But so far I haven't even begun testing this and now I need to focus on the missing token events.

**Comment by jsc-paneda** [comment]
> > However!? I get much more audible dropouts with 4.4, roughly one audible glitch every 10 sec.
> > With the 4.19 kernel I only get audible glitches when I also detect a missing event.
> 
> Can you check what's the cpu usage of 4.4? Maybe with another system you are doing different things (for instance, mpd may do format/sample rate conversion and throttle the CPU) Please make sure you are using the same MPD/ALSA setup and the same audio file.

CPU load average for all tests are _very_ close to zero (~0.02). I do use the Alsa 'plug' plugin however my sound device handles the specified samplerate so I expect Alsa not to perform any src.
I only use the arecord .. | aplay as specified above, i.e. no MPD daemons are running..

**Comment by alastair-dm** [comment]
I've reproduced audible pops now with Pi 3 B, Raspbian Lite, squeezelite and brutefir with a brutefir config based on the one i had in Volumio (device names and paths changed to work in Raspbian). I've tried with the stock kernel, a self-compiled 4.19.25-v7+ from a checkout of the rpi-4.19.y branch, and the same with hand-reversion of the changes I suggested earlier. The card is the same, with the kernel chosen in config.txt. Network is wired. With all kernels the audible pops are subjectively similar, but I don't have an objective measure. I get occasional pops (10s of seconds apart usually) streaming 128kbps mp3, and more frequent pops streaming dxd flac files originally downloaded from [2L's samples](http://www.2l.no/hires/index.html). Subjectively the pops seem worse than with volumio/brutefir on the same hardware, but it's hard to be sure. Is there anything in the kernel I should enable to get an objective measure of the glitches?

**Comment by jsc-paneda** [comment]
Got some new interesting results now.

Result 1. 
I found that the device USB library also has a method for getting SOF events. I then realized that the data methods I've been using for measuring duration between successive calls really measures 'TOKEN' events. So my terminology has been off up until now.
In my audio device I then added monitors for detecting missing events for all three types (in/out tokens and sof events). 
I can now confirm that _no_ SOF events can be detected as missing. However 'TOKEN' events are the ones that I detect missing (see my previous posts). I will edit my previous posts so that they use the correct terminology.

Result 2.
I have spent lots of time building kernels and swapped in kernels of different versions on the same sd card. I have not found any particular kernel version to perform better or worse (at least not with confident results). 
However in general I found that ArchLinux seems to loose lots more token events than Raspian with the same kernel version. As ArchLinux and Raspian differs alot regarding their kernel configuration, I've tried to match section by section to see if any particular kernel config section affects the behaviour. However so far I found none. So at the moment I have given up on this work. However the difference between the two distros are quit big (note that comparison is somewhat insecure as two different sd cards was used). 

Result 3 (most interesting).
Up until now I have mainly used clean ArchLinux/Raspian builds with the 'arecord|aplay' command. The cpu load has been close to zero. I then thought to test would would happen if I put the CPU under more stress. So I used the tool [stress](http://people.seas.harvard.edu/~apw/stress/).

To my surprise when I run 'stress -c3' (putting 3 cores at a 100% cpu load), I suddenly stopped loosing 'token' events! This was very much the opposite of what I expected.

I then thought that perhaps the varying cpu frequencies (default the frequency jump between 600MHz and 1.2GHz on RPi 3) affects the result. So I added 'force_turbo=1' to /boot/config.txt. This keep the frequency fixed at 1.2GHz. ~~This did not help, I still detect missing token events.~~ This does have an impact. I still detect missing token events but at a much lower rate.

However whenever I run 'stress -c 3', I loose no token events and the audio is glitch free. Perhaps some other Linux scheduling kicks in or something...

**Comment by jsc-paneda** [comment]
I have now managed to work around the problem of missing token events by changes in the device firmware.
This does obviously not help to solve the real problem for others but it still may be good info.

Workaround:
My particular device was using asynchronous mode with implicit feedback (a mode of isochronous transfer). I changed this to asynchronous mode with explicit feedback. 
Through the explicit feedback the host (Alsa/snd_usb_audio) gets information about the rate it should use when forwarding samples to the device. Typically the device calculates this rate by comparing sof events with the on board master clock (chap. 5.12.4.2 in the USB 2 specification). The same rate factor is also used internally in the device to send samples to the host. 
However since we're loosing token events, this rate factor cannot keep the device buffer from underflow/overflow so the workaround is simply to add a check of the device buffers for in/out. If they are about to underflow/overflow, the rate factor is overridden for both directions. This guarantees that the device buffer cannot underflow/overflow due to missing token events.

Result:
I have used different versions of the kernel 4.4, 4.14, 4.19. Regardless of kernel version, _no_ audible dropouts can be detect anymore.

So at least for UAC2 devices using asynchronous mode/explicit feedback, the missing token event issue is the _only_ issue for audible dropouts.

Its however important to note that other devices that uses other modes (adaptive/synchronous), there might be other causes for audible dropouts. However my guess is that the missing token events accounts for the dropouts also for these modes.

**Comment by alastair-dm** [comment]
In light of @jsc-paneda's findings using 'stress -c3' I tried running with only one core enabled by appending maxcpus=1 to /boot/cmdline.txt then repeating my tests with brutefir. I didn't hear any glitches, but without an objective test I can't be certain they were completely cured. CPU use was ~30% with brutefir outputting 96kHz s32_le on 2 channels, and squeezelite streaming 24/352 flac dxd sample files.

**Comment by alastair-dm** [comment]
After some further testing it seems I only need to pin brutefir to one cpu core to avoid glitches (or at least make them rare enough that I'm not certain they're happening - I really need an objective test to find them.) This can be done using taskset when starting brutefir, such as:
sudo taskset -c 1 brutefir /home/pi/brutefir.cfg &
I got to this stage by first switching from 'maxcpus=1' to 'isolcpus=1,2,3' which was still glitch free without using taskset. I then tried giving brutefir and squeezelite access to cores 1-3 with 'taskset -c 1-3' and the glitches returned. Only when brutefir was restricted to a single cpu did the glitches stop again. I then removed the isolcpus parameter from the command line and kept taskset for brutefir only, and it seems glitch free.

**Comment by aas72** [comment]
After setting the following:
dwc_otg.nak_holdoff_enable=0

Using volumio with drc my clicks and pop were gone. I'm still interested in an explanation from some expert. I hope this is helpful to other raspberry/xmos (usb dac) users.

**Comment by nL1ght5** [comment]
The same issue is present running MPD in combination with DAC using Amanero Combo 384 module as USB interface. I discovered that a continuous iperf measurement adding traffic over the physical network interface does a good job of provoking glitches. Stressing CPUs removes the glitches as other users have mentioned, Isolating MPD to single CPU does not resolve the issue.

**Comment by Phasco** [comment]
I also see this issue with a MiniDSP USB UDAC8 connected to a Raspberry Pi 3B+ running BruteFIR via ALSA. There is a null drop-out every 1-2 seconds which increases with background network activity.
It's not BruteFIR as 6 filters of 16384 taps at 44100 Hz are using only 17% average CPU across 4 cores.
If I output the 6 audio streams via HDMI (PCM) the problem goes away. I therefore conclude it's a latency problem with the USB output device. Either the buffers are too small or some other device is taking huge gulps of non-preemptive USB time.
(I'm using Raspian Stretch )

**Comment by popcornmix** [comment]
> Stressing CPUs removes the glitches as other users have mentioned

does force_turbo=1 have the same effect without stressing?

**Comment by nL1ght5** [comment]
force_turbo=1 may affect the outcome, however there is still glitches when there is network activity of the physical network interface.

**Comment by Phasco** [comment]
> > Stressing CPUs removes the glitches as other users have mentioned
> 
> does force_turbo=1 have the same effect without stressing?

Hi popcornmix - love your work.

To answer your question - No, force_turbo=1 does not have the same effect as stressing.
In fact, based on the random nature of the glitches, I would guess force_turbo has no effect.

My testing:
Using raspian stretch 4.14.98-v7+
I'm not using the ethernet interface - all network via onboard WIFI
Testing with a 1KHz sine wave from REW to USB DAC direct UDAC8 [plughw:2,0].
(All the intervals below are averages - individual glitches are random and can be very close together.)

force_turbo=0
At idle: glitches more than once per second
stress -c 1: glitches average one - two seconds
stress -c 2: glitches about every three seconds
stress -c 3: one glitch in one minute (stopped listening)

Then restarted with force_turbo=1
At idle: glitches about every two seconds
stress -c 1: glitches about every five seconds
stress -c 2: glitches about every five seconds
stress -c 3: two glitches in one minute (stopped listening)

I then did the following to try and stop the glitches (With force_turbo at idle):
 - disconnected the WIFI network interface
 - killed CRON
 - killed avahi
 - I then continued to kill every task I could until the system died. Glitches remained until then end.
 - Note that as I scrolled up and down in htop the glitches are more frequent.

I'm happy to test other scenarios if you like - I really want to help solve this as I can't get multichannel audio working out of the hdmi interface under alsa - so I'm forced to use a multichannel USB DAC.

**Comment by Wang-Yue** [comment]
RPi 4 is released today. 
It seems they improved the USB speed significantly (now they have usb-c and usb 3.0).
 I'm waiting to see if the issue is solved by the hardware upgrade.
Anyone ordered the new board?

**Comment by ghost** [comment]
The USB interface on the Pi 4 is completely different to all previous Pi's. On the Pi 4 USB is handled by a separate chip connected to the SoC via a PCI Express bus. The SoCs own USB 2.0 interface, which is the same as on the BCM283x's, is wired to the USB-C power connector for use as an OTG port.

I would therefore expect that the Pi 4 is immune to almost all USB problems that affected previous models.

**Comment by JamesH65** [comment]
That is the hope!!!! Completely different USB controller equals completely different set of bugs..!!

**Comment by alastair-dm** [comment]
Do you have any links for documentation of the OTG connection on the USB-C? I didn't spot it in anything I found earlier today. Having both the OTG and PCI-e connected USB-2 may be a helpful diagnostic.

**Comment by ghost** [comment]
> That is the hope!!!! Completely different USB controller equals completely different set of bugs..!!

Indeed, but given that the new USB controller is much more suited to the job of interfacing with multiple devices simultaneously, I would not expect there to be any major issues. Certainly they should be easier to fix. Just need to avoid bugs in the hardware 😮

**Comment by volumio** [comment]
In any case, I am still very keen to find a solution for previous generation's PI.
I am able to get glitch free playback on any format, except DSD256 in DoP mode. 

My best conf is the following (in cmdline.txt):
`dwc_otg.fiq_enable=1 dwc_otg.fiq_fsm_enable=1 dwc_otg.fiq_fsm_mask=0xF dwc_otg.nak_holdoff=1`

However, I still have glitches (with some DACs) in DSD256 DoP. Any hint?

**Comment by spockfish** [comment]
Even with the mask=0xf I have occasional glitches. However, I can get 'glitch free' completely by forcing the CPU on a constant speed.

It seems that switching between 2 freq's (which happens normally when I play audio) has influence on this.

**Comment by lategoodbye** [comment]
Might be interesting to compare the behavior of [Downstream](https://github.com/raspberrypi/linux/blob/rpi-4.19.y/drivers/cpufreq/bcm2835-cpufreq.c) and [Mainline cpufreq](https://github.com/raspberrypi/linux/blob/rpi-5.3.y/drivers/cpufreq/raspberrypi-cpufreq.c) driver according to this issue.

**Comment by DBMandrake** [comment]
Despite valiant attempts at fixes in the kernel my feeling is that this is a hardware architectural limtitation in the original Pi models (1 to 3) that cannot be fixed in software, and two years after reporting this issue it remains unresolved.

Has anyone done any thorough testing of this issue with an affeced USB sound interface on the Raspberry Pi 4 ? If this problem does indeed not exist on the Pi 4 with its different USB controller architecture then I'm willing to give in and simply buy a Pi 4 for use with Volumio.

**Comment by alastair-dm** [comment]
I tried to reproduce the problem on the Pi 4 using both the type A sockets that use the new controller, and the type C socket that uses the old one. I couldn't reproduce it in either case using the same configuration I used to test on the Pi 3 - squeezelite -> BruteFIR -> Focusrite Forte streaming high res files over wired ethernet. I had intended to repeat the tests on a Pi 3 to check that it wasn't a software change that had sorted it since there's also been a major raspbian version change, but I haven't done so yet. i should also try it using the Pi Zero W I suppose, in case the problem is related to the usb hub rather than the controller.

**Comment by DBMandrake** [comment]
That sounds promising... :+1:

**Comment by pelwell** [comment]
Pi 4 has a full XHCI-compliant USB3 controller with far more throughput than the DWC-OTG controller on previous Pis, and it puts very little load on the ARMs. I'm confident that it would solve your audio dropouts.

**Comment by spockfish** [comment]
Agreed @pelwell , but that still does not explain why I've got drop outs pops and glitches with a 4.19 kernel on a Pi 3, and if I revert to 4.14 they are gone.

**Comment by AlloKatana** [comment]
We have designed an audio streamer based on CM 3+ and we see the same issue .

  We have lots of reports of glitches with 4.19 and not with 4.14

Is anything done on this issue ?

**Comment by pelwell** [comment]
Something clearly changed in scheduling or interrupt handling between 4.14 and 4.19. I'm sure the change would be common to all platforms, but Pis 0 to 3 are probably more sensitive to it because of the lack of latency tolerance in the OTG USB controller.

**Comment by DBMandrake** [comment]
@AlloKatana and others,

Let's try not to muddy the waters too much - the problem I reported in this git issue is not a 4.19 regression and occurs on all earlier Raspberry Pi kernel versions I've tested including 4.14.

Not all USB audio glitches and dropouts are the same issue I reported, and while there may well be a regression in 4.19 which has made things much worse it should probably be discussed in its own separate Git issue specific to 4.19 regressions.

**Comment by volumio** [comment]
We have contacted PI foundation to find a proper solution to this. The proposed solution (still to be implemented) is to add another bit in the fiq_fsm_mask parameter that pipelines isochronous transfers.
See: https://www.raspberrypi.org/forums/viewtopic.php?f=98&t=239388

Unfortunately, we've not heard back from foundation yet about this. When there will be any progress I'll post it here.

**Comment by DBMandrake** [comment]
@volumio Thanks for the pointer to that thread, I'll keep an eye on it. Just yesterday I've ordered a Pi 4 to be my new Volumio box, so for me at least the problem should be solved soon, (and I will no longer need to rely on a Cubox-i running very old Volumio builds) however it would still be great of course if a workable solution can be found for earlier models of Pi.

My only concern is that the symptoms can be subtle and very hard to reproduce consistently - I've tried many different cmdline.txt and config.txt options (on a Pi 2) people have said work for them including all the ones suggested here and sometimes at first it seems to work but after using it for longer and trying different music and/or sample rates it becomes apparent that the problem is still there but perhaps just a bit less frequent and more difficult to trigger. So I suspect the people reporting success with various settings just aren't testing long enough or thoroughly enough on a wide enough variety of media files.

Hopefully any fiq_fsm_mask fix doesn't end up just making it occur very infrequently instead of actually fixing it fully.

**Comment by DBMandrake** [comment]
I received my Pi 4 a few days ago and I can confirm that the issues I reported here are 100% resolved with the Pi 4. :smile: Perfect glitch free playback in Volumio up to 192Khz 24bit using the same Behringer UMC204HD which was having untold problems with dropouts on the Pi 2.

An unexpected side bonus is that the 5 volt output of the Pi 4 seems to be **infinitely** cleaner than that of the Pi 2.

My Pi 2 was using the official (at the time) Pi 2 power adaptor and yet the USB 5 volt output was so noisy that "CPU chirping" and other digital noise was readily apparent on the analogue output of the DAC with music paused even at normal amplifier volume settings. Connecting a TV to the HMDI output (normally disconnected with Volumio) made the noise even worse, almost intolerable.

On the Pi 4 (also with the official USB-C power adaptor) there is no CPU chirping or digital noise audible at all. I can turn the amplifier way up to 2 o'clock (well beyond even loud playback) while paused and only hear a small amount of smooth analogue hiss from the amplifier itself that is present with the DAC unplugged.

Basically no power noise of any kind is passing from the Pi through the DAC at all. I'm stunned.

I don't know whether the improvement is coming from the power adaptor or from filtering of the 5v supply to the USB ports on the Pi 4 PCB - any insights into the analogue/power design of the Pi 4 which might explain this dramatic reduction in USB 5v noise @pelwell or @P33M ?

**Comment by pelwell** [comment]
Thanks for the feedback - if we ever need some poster quotes, we'll know where to come.

I'm not going to embarrass myself with theories about where the improvements come from, but @jimbojr goes to a lot of effort over details like this.

**Comment by vedranius** [comment]
Thing is, on Raspberry Pi 3B+ and Raspberry Pi 4 4GB, Pi 400 and 8GB I have same problems.
Running Patchbox OS which is actually an distro based on Raspbian. 
Im testing it with Focusrite Scarlett 8i6.
Tried all kinds of sample rates and whichever I try I get crackling and popping sound, sometime momentarilly, sometimes after 10 minutes or even 1 to 2 hours. 
Im looking hard for a solution but so far I cant find anything that might cause this problem.
First I was thinking it was a power supply, but Im using official adapters from RPi, Ive also tried connecting them to Anker professional powerbanks and nothing changed. 
Also, tried jackd versions 1.9.11 and 1.9.12 tho Id like to update to the latest jackd 1.9.16 but not sure how to update and safely remove previous version, I just couldnt find a way... 

Did anyone had similar problem? 
P. S. Ive done tests also on Behringer UMC202, ESI U24 XL and some other usb audio interfaces. 
The only that works fine is PiSound which is connected directly via GPIO... 
Thanks!

**Comment by skateskate** [comment]
Any updates on this issue? I am having the same problems with a Raspberry Pi Zero W and an USB XMOS DAC.

**Comment by silvestropomes** [comment]
Any news?
I have an SMSL SU-8 DAC (with an XMOS X208 USB receiver) and I've the same problem.
With the RPi2 I had frequent drops (a lot per minute), with the RPi4 it's much much better, as I thought the problem was solved, but it wasn't.
I've glitches when there's an high network (or maybe CPU also?) usage: when I use the UI, (loading artists from the NAS, browsing a DLNA source...) and also when simply Volumio uses the network to fill the internal buffer. It depends, once every 20-30 seconds? It depends to the buffer size and the music bitrate.

I really hoped RPi4 solved the problem... but it went just better, not fine.

**Comment by nyebi** [comment]
> Any news?
> I have an SMSL SU-8 DAC (with an XMOS X208 USB receiver) and I've the same problem.
> With the RPi2 I had frequent drops (a lot per minute), with the RPi4 it's much much better, as I thought the problem was solved, but it wasn't.
> I've glitches when there's an high network (or maybe CPU also?) usage: when I use the UI, (loading artists from the NAS, browsing a DLNA source...) and also when simply Volumio uses the network to fill the internal buffer. It depends, once every 20-30 seconds? It depends to the buffer size and the music bitrate.
> 
> I really hoped RPi4 solved the problem... but it went just better, not fine.

I have different problems with the RPi4 USB-A ports: my USB DAC can only output 2 channel audio because if i try to use it in 5.1 mode i only get silence and an error message about not enough bandwidth on the bus for the device. Afaik it's a problem with the integrated USB 3.0 HUB. However i found that if i set the Pi4's USB-C port to host mode and connect the DAC to it using a USB-C otg cable it works perfectly fine.

**Comment by DBMandrake** [comment]
> Any updates on this issue? I am having the same problems with a Raspberry Pi Zero W and an USB XMOS DAC.

Three years and no fix is forthcoming - I don't think this is fixable in software, it is a hardware limitation in the USB controller, so waiting for a software fix at this point I think is wishful thinking.

I gave up long ago and switched to a Pi 4 which has a completely different USB controller which doesn't have the same limitations.

> Any news?
> I have an SMSL SU-8 DAC (with an XMOS X208 USB receiver) and I've the same problem.
> With the RPi2 I had frequent drops (a lot per minute), with the RPi4 it's much much better, as I thought the problem was solved, but it wasn't.
> I've glitches when there's an high network (or maybe CPU also?) usage: when I use the UI, (loading artists from the NAS, browsing a DLNA source...) and also when simply Volumio uses the network to fill the internal buffer. It depends, once every 20-30 seconds? It depends to the buffer size and the music bitrate.
> 
> I really hoped RPi4 solved the problem... but it went just better, not fine.

I've been using a 1GB Pi 4 with Volumio for over a year now with the same USB DAC from my initial post and it is 100% glitch free for me at all sample/bitrates right up to 192Khz / 24bit. Doesn't matter whether I am playing back local content or over the network. Absolutely zero problems.

I'm not sure what might be causing your issue - are you using Ethernet or WiFi ? (I use Ethernet)

Have you tried plugging it into both USB 2 and USB 3 ports? (I use the USB 2 ports although I haven't noticed any problems on the USB 3 ports - worth trying both though)

Does it only happen at higher sample / bitrates or even down at 44Khz / 16 bit? Is it any worse or better at different sample/bitrates ?

Perhaps it is something specific to your sound card...

A good test is to generate test files at each sample / bitrate with a 20Hz sinewave - during playback if there are any dropouts/glitches they will be audible as a slight click/pop because the 20Hz sinewave itself will not be audible on most speakers...(or if it is, turn the bass down during the test...) If there are no pops or clicks and you just hear silence after playback starts there is no issue.

**Comment by P33M** [comment]
Wontfix - the dwc_otg codebase is effectively in maintenance mode. High-speed isochronous devices with a bInterval of 1 are supported on a best-effort basis and to increase compatibility with these means meddling with FIQ code in a non-trivial manner.

dwc2 in host mode (as used on Pi Zero / Pi Zero W) will suffer from the same root problem as dwc_otg - the recommendation is to use full-speed mode in this case.

**Comment by DBMandrake** [comment]
Thanks for the final report on this problem - I had guessed this issue would not be solvable due to hardware limitations and I've long since moved on to using a Pi 4 in this application, which works perfectly. I suggest anyone else with the same sort of USB DAC glitches/dropout described here does the same.

**Comment by larschristensen** [comment]
Unfortunately, it appears moving to Raspberry Pi4 doesn't solve all the USB-related audio problems. For this reason, I have given up on using Raspberry Pi4 for this purpose. For people still experiencing USB-related audio problems, I can recommend Odroid boards for ARM-based development or simply going with x86-based Linux.

**Comment by fccoelho** [comment]
FWIW, I would like to report  very similar symptoms: audio dropouts every few seconds on a UMC204HD playback on a much more powerful  machine: Ryzen 9 with 32GB of RAM and  16 threads. So this is not a problem of Raspberry Pi not handling the load, it's probably an issue of the Audio interface.

**Comment by nyebi** [comment]
If anyone is interested i found a workaround: if i supply power to an Rpi4 through the GPIO header and use the USB-C port through an USB-A -> USB-C converter as host to connect the DAC it works fine. No dropouts and 5.1 audio works fine too.
